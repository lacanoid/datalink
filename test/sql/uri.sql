\pset null _null_

SET client_min_messages = notice;
SET search_path = public,datalink;

alter table sample_urls alter url type datalink.dl_url;

insert into sample_urls (url)
values 
  ('file://user@fs1.arch.net/file1.jpg'),
  ('file://user@file-server/a/b/c/../../fileX.jpg'),
  ('file://user@fs2.arch.net/file1.jpg#bb519e69-6bf1-4214-9b7a-4280702c6321'),
  ('file://user:pass@host/fileA.txt'),
  ('ftp://user:pass@host/fileB.txt'),
  ('sftp://user:pass@host/fileC.txt#80653c50-908b-4f74-a7ca-398a6035a2d1'),
  ('mailto:lacanoid@ljudmila.org?subject=feedback'),
  ('https://video.kulturnik.si/?qwhat=galerije&q=piran'),
  ('file:///var/www/datalink/'),
  ('file:///var/www/datalink')
;

insert into sample_urls (url) values ('data:image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs=');

select url,
       uri_get(url,'scheme') as scheme,
       uri_get(url,'userinfo') as userinfo,
       uri_get(url,'host') as host,
       uri_get(url,'path') as path,
       uri_get(url,'query') as query,
       uri_get(url,'fragment') as fragment,
       uri_get(url,'only') as only
  from sample_urls
;

select url,
       uri_get(url,'scheme') as scheme,
       uri_get(url,'server') as server,
       uri_get(url,'basename') as basename,
       uri_set(url,'basename','foo.bar')
  from sample_urls
;

select url,dlpreviouscopy(url::text,0) from sample_urls;

select 'foo/bar'::datalink.file_path;
select '/foo/bar'::datalink.file_path;
select '/foo/bar/..'::datalink.file_path;
select '/foo/../bar'::datalink.file_path;

select dlvalue('https://www.github.org');
select dlvalue('https://www.github.org')::uri;

\x
with d as (
 select 'file:///var/www/datalink/test4.txt#krneki' as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;

with d as (
 select dlvalue('file:///var/www/datalink/test4.txt#krneki') as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;

with d as (
 select 'file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced' as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;

with d as (
 select dlvalue('file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced') as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;

with d as (
 select dlpreviouscopy(dlvalue('file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced')) as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link)
  from d;

;
\x
