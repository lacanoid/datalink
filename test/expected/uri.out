\pset null _null_
SET client_min_messages = notice;
SET search_path = public,datalink;
-- test dirurl mapping
select dlvalue('http://localhost/datalink/test1.txt');
                   dlvalue                   
---------------------------------------------
 {"a": "file:///var/www/datalink/test1.txt"}
(1 row)

select dlvalue(datalink.uri_set((dlvalue('http://localhost/datalink/test1.txt')::jsonb->>'a')::uri,'src','index.rtf'),null,'hello');
                          dlvalue                          
-----------------------------------------------------------
 {"a": "file:///var/www/datalink/index.rtf", "c": "hello"}
(1 row)

select dlvalue('index.html',dlvalue('http://localhost/datalink/test1.txt'));
                   dlvalue                    
----------------------------------------------
 {"a": "file:///var/www/datalink/index.html"}
(1 row)

select dlvalue('',dlvalue('http://localhost/datalink/test1.txt'));
                   dlvalue                   
---------------------------------------------
 {"a": "file:///var/www/datalink/test1.txt"}
(1 row)

select dlvalue(null,dlvalue('http://localhost/datalink/test1.txt'));
                   dlvalue                   
---------------------------------------------
 {"a": "file:///var/www/datalink/test1.txt"}
(1 row)

alter table sample_urls alter url type datalink.url;
insert into sample_urls (url)
values 
  ('file://user@fs1.arch.net/file1.jpg'),
  ('file://user@file-server/a/b/c/../../fileX.jpg'),
  ('file://user@fs2.arch.net/file1.jpg#bb519e69-6bf1-4214-9b7a-4280702c6321'),
  ('file://user:pass@host/fileA.txt'),
  ('ftp://user:pass@host/fileB.txt'),
  ('sftp://user:pass@host/fileC.txt#80653c50-908b-4f74-a7ca-398a6035a2d1'),
  ('mailto:lacanoid@ljudmila.org?subject=feedback'),
  ('https://video.kulturnik.si/?qwhat=galerije&q=piran'),
  ('file:///var/www/datalink/'),
  ('file:///var/www/datalink')
;
insert into sample_urls (url) values ('data:image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs=');
select url,
       uri_get(url,'scheme') as scheme,
       uri_get(url,'userinfo') as userinfo,
       uri_get(url,'host') as host,
       uri_get(url,'path') as path,
       uri_get(url,'query') as query,
       uri_get(url,'fragment') as fragment,
       uri_get(url,'only') as only
  from sample_urls
;
                                   url                                   | scheme | userinfo  |        host        |                     path                      |         query          |               fragment               |                        only                        
-------------------------------------------------------------------------+--------+-----------+--------------------+-----------------------------------------------+------------------------+--------------------------------------+----------------------------------------------------
 http://www.mozilla.org                                                  | http   | _null_    | www.mozilla.org    |                                               | _null_                 | _null_                               | http://www.mozilla.org
 http://www.ljudmila.org                                                 | http   | _null_    | www.ljudmila.org   |                                               | _null_                 | _null_                               | http://www.ljudmila.org
 https://www.github.org                                                  | https  | _null_    | www.github.org     |                                               | _null_                 | _null_                               | https://www.github.org
 file:///tmp/a                                                           | file   | _null_    |                    | /tmp/a                                        | _null_                 | _null_                               | file:///tmp/a
 http://www.debian.org/                                                  | http   | _null_    | www.debian.org     | /                                             | _null_                 | _null_                               | http://www.debian.org/
 file://user@fs1.arch.net/file1.jpg                                      | file   | user      | fs1.arch.net       | /file1.jpg                                    | _null_                 | _null_                               | file://user@fs1.arch.net/file1.jpg
 file://user@file-server/a/b/c/../../fileX.jpg                           | file   | user      | file-server        | /a/b/c/../../fileX.jpg                        | _null_                 | _null_                               | file://user@file-server/a/fileX.jpg
 file://user@fs2.arch.net/file1.jpg#bb519e69-6bf1-4214-9b7a-4280702c6321 | file   | user      | fs2.arch.net       | /file1.jpg                                    | _null_                 | bb519e69-6bf1-4214-9b7a-4280702c6321 | file://user@fs2.arch.net/file1.jpg
 file://user:pass@host/fileA.txt                                         | file   | user:pass | host               | /fileA.txt                                    | _null_                 | _null_                               | file://user:pass@host/fileA.txt
 ftp://user:pass@host/fileB.txt                                          | ftp    | user:pass | host               | /fileB.txt                                    | _null_                 | _null_                               | ftp://user:pass@host/fileB.txt
 sftp://user:pass@host/fileC.txt#80653c50-908b-4f74-a7ca-398a6035a2d1    | sftp   | user:pass | host               | /fileC.txt                                    | _null_                 | 80653c50-908b-4f74-a7ca-398a6035a2d1 | sftp://user:pass@host/fileC.txt
 mailto:lacanoid@ljudmila.org?subject=feedback                           | mailto | _null_    | _null_             | lacanoid@ljudmila.org                         | subject=feedback       | _null_                               | mailto:lacanoid@ljudmila.org?subject=feedback
 https://video.kulturnik.si/?qwhat=galerije&q=piran                      | https  | _null_    | video.kulturnik.si | /                                             | qwhat=galerije&q=piran | _null_                               | https://video.kulturnik.si/?qwhat=galerije&q=piran
 file:///var/www/datalink/                                               | file   | _null_    |                    | /var/www/datalink/                            | _null_                 | _null_                               | file:///var/www/datalink/
 file:///var/www/datalink                                                | file   | _null_    |                    | /var/www/datalink                             | _null_                 | _null_                               | file:///var/www/datalink
 data:image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs=                      | data   | _null_    | _null_             | image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs= | _null_                 | _null_                               | data:image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs=
(16 rows)

select url,
       uri_get(url,'scheme') as scheme,
       uri_get(url,'server') as server,
       uri_get(url,'basename') as basename,
       uri_set(url,'basename','foo.bar')
  from sample_urls
;
                                   url                                   | scheme |       server       |                basename                 |                                uri_set                                
-------------------------------------------------------------------------+--------+--------------------+-----------------------------------------+-----------------------------------------------------------------------
 http://www.mozilla.org                                                  | http   | www.mozilla.org    | _null_                                  | http://www.mozilla.org/foo.bar
 http://www.ljudmila.org                                                 | http   | www.ljudmila.org   | _null_                                  | http://www.ljudmila.org/foo.bar
 https://www.github.org                                                  | https  | www.github.org     | _null_                                  | https://www.github.org/foo.bar
 file:///tmp/a                                                           | file   |                    | a                                       | file:///tmp/foo.bar
 http://www.debian.org/                                                  | http   | www.debian.org     | _null_                                  | http://www.debian.org/foo.bar
 file://user@fs1.arch.net/file1.jpg                                      | file   | fs1.arch.net       | file1.jpg                               | file://user@fs1.arch.net/foo.bar
 file://user@file-server/a/b/c/../../fileX.jpg                           | file   | file-server        | fileX.jpg                               | file://user@file-server/a/b/c/../../foo.bar
 file://user@fs2.arch.net/file1.jpg#bb519e69-6bf1-4214-9b7a-4280702c6321 | file   | fs2.arch.net       | file1.jpg                               | file://user@fs2.arch.net/foo.bar#bb519e69-6bf1-4214-9b7a-4280702c6321
 file://user:pass@host/fileA.txt                                         | file   | host               | fileA.txt                               | file://user:pass@host/foo.bar
 ftp://user:pass@host/fileB.txt                                          | ftp    | host               | fileB.txt                               | ftp://user:pass@host/foo.bar
 sftp://user:pass@host/fileC.txt#80653c50-908b-4f74-a7ca-398a6035a2d1    | sftp   | host               | fileC.txt                               | sftp://user:pass@host/foo.bar#80653c50-908b-4f74-a7ca-398a6035a2d1
 mailto:lacanoid@ljudmila.org?subject=feedback                           | mailto | _null_             | lacanoid@ljudmila.org                   | mailto:lacanoid@ljudmila.org?subject=feedback/foo.bar
 https://video.kulturnik.si/?qwhat=galerije&q=piran                      | https  | video.kulturnik.si | _null_                                  | https://video.kulturnik.si/foo.bar?qwhat=galerije&q=piran
 file:///var/www/datalink/                                               | file   |                    | _null_                                  | file:///var/www/datalink/foo.bar
 file:///var/www/datalink                                                | file   |                    | datalink                                | file:///var/www/foo.bar
 data:image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs=                      | data   | _null_             | gif;base64,R0lGODlhyAAiALM...DfD0QAADs= | data:image/foo.bar
(16 rows)

select url,dlpreviouscopy(url::text,0) from sample_urls;
                                   url                                   |                                       dlpreviouscopy                                       
-------------------------------------------------------------------------+--------------------------------------------------------------------------------------------
 http://www.mozilla.org                                                  | {"a": "http://www.mozilla.org", "k": "p"}
 http://www.ljudmila.org                                                 | {"a": "http://www.ljudmila.org", "k": "p"}
 https://www.github.org                                                  | {"a": "https://www.github.org", "k": "p"}
 file:///tmp/a                                                           | {"a": "file:///tmp/a", "k": "p"}
 http://www.debian.org/                                                  | {"a": "http://www.debian.org/", "k": "p"}
 file://user@fs1.arch.net/file1.jpg                                      | {"a": "file://user@fs1.arch.net/file1.jpg", "k": "p"}
 file://user@file-server/a/b/c/../../fileX.jpg                           | {"a": "file://user@file-server/a/fileX.jpg", "k": "p"}
 file://user@fs2.arch.net/file1.jpg#bb519e69-6bf1-4214-9b7a-4280702c6321 | {"a": "file://user@fs2.arch.net/file1.jpg#bb519e69-6bf1-4214-9b7a-4280702c6321", "k": "p"}
 file://user:pass@host/fileA.txt                                         | {"a": "file://user:pass@host/fileA.txt", "k": "p"}
 ftp://user:pass@host/fileB.txt                                          | {"a": "ftp://user:pass@host/fileB.txt", "k": "p"}
 sftp://user:pass@host/fileC.txt#80653c50-908b-4f74-a7ca-398a6035a2d1    | {"a": "sftp://user:pass@host/fileC.txt#80653c50-908b-4f74-a7ca-398a6035a2d1", "k": "p"}
 mailto:lacanoid@ljudmila.org?subject=feedback                           | {"a": "mailto:lacanoid@ljudmila.org?subject=feedback", "k": "p"}
 https://video.kulturnik.si/?qwhat=galerije&q=piran                      | {"a": "https://video.kulturnik.si/?qwhat=galerije&q=piran", "k": "p"}
 file:///var/www/datalink/                                               | {"a": "file:///var/www/datalink/", "k": "p"}
 file:///var/www/datalink                                                | {"a": "file:///var/www/datalink", "k": "p"}
 data:image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs=                      | {"a": "data:image/gif;base64,R0lGODlhyAAiALM...DfD0QAADs=", "k": "p"}
(16 rows)

select 'foo/bar'::datalink.file_path;
ERROR:  value for domain file_path violates check constraint "file_path_absolute"
select '/foo/bar'::datalink.file_path;
 file_path 
-----------
 /foo/bar
(1 row)

select '/foo/bar/..'::datalink.file_path;
ERROR:  value for domain file_path violates check constraint "file_path_noparent"
select '/foo/../bar'::datalink.file_path;
ERROR:  value for domain file_path violates check constraint "file_path_noparent"
select dlvalue('https://www.github.org');
             dlvalue             
---------------------------------
 {"a": "https://www.github.org"}
(1 row)

select dlvalue('https://www.github.org')::uri;
        dlvalue         
------------------------
 https://www.github.org
(1 row)

\x
with d as (
 select 'file:///var/www/datalink/test4.txt#krneki' as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;
-[ RECORD 1 ]-----+-------------------------------------------
link              | file:///var/www/datalink/test4.txt#krneki
dlurlcomplete     | http://localhost/datalink/test4.txt#krneki
dlurlcompleteonly | http://localhost/datalink/test4.txt
dlurlpath         | /var/www/datalink/test4.txt
dlurlpathonly     | /var/www/datalink/test4.txt

with d as (
 select dlvalue('file:///var/www/datalink/test4.txt#krneki') as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;
-[ RECORD 1 ]-----+---------------------------------------------------
link              | {"a": "file:///var/www/datalink/test4.txt#krneki"}
dlurlcomplete     | http://localhost/datalink/test4.txt#krneki
dlurlcompleteonly | http://localhost/datalink/test4.txt
dlurlpath         | /var/www/datalink/test4.txt
dlurlpathonly     | /var/www/datalink/test4.txt

with d as (
 select 'file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced' as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;
-[ RECORD 1 ]-----+-------------------------------------------------------------------------
link              | file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced
dlurlcomplete     | http://localhost/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced
dlurlcompleteonly | http://localhost/datalink/test3.txt
dlurlpath         | /var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced
dlurlpathonly     | /var/www/datalink/test3.txt

with d as (
 select dlvalue('file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced') as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       dlurlpath(link),
       dlurlpathonly(link)
  from d;
-[ RECORD 1 ]-----+---------------------------------------------------------------------------------
link              | {"a": "file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced"}
dlurlcomplete     | http://localhost/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced
dlurlcompleteonly | http://localhost/datalink/test3.txt
dlurlpath         | /var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced
dlurlpathonly     | /var/www/datalink/test3.txt

with d as (
 select dlpreviouscopy(dlvalue('file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced')) as link
)
select link,
       dlurlcomplete(link),
       dlurlcompleteonly(link),
       datalink.uri_get(link,'basename') as basename
  from d;
-[ RECORD 1 ]-----+---------------------------------------------------------------------------------------------------
link              | {"a": "file:///var/www/datalink/test3.txt", "b": "11111111-2222-3333-4444-abecedabeced", "k": "p"}
dlurlcomplete     | http://localhost/datalink/test3.txt
dlurlcompleteonly | http://localhost/datalink/test3.txt
basename          | test3.txt

;
\x
